django_pg_simple_hll/migrations/0001_initial.sql

```diff
@@ -0,0 +1,161 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
+-- hll_precision is the precision we are using for the approximation
+--  it use used to calculate the number of buckets
+create or replace function hll_bucket(
+    hll_agg_state int[],
+    input anyelement,
+    hll_precision int
+) returns int[]
+language plpgsql immutable strict parallel safe as $$
+declare
+    -- eg. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    n_buckets int := POW(2, hll_precision);
+    -- transform the input into a text and hash it
+    hashed_input int := hashtext(input::text);
```

Comment:
So `hashtext` is an undocumented internal postgres function: are there any particular properties we need of the hash function for this aggregate to work correctly?

Reply:
Yes. It needs to produce uniformly distributed integers from the input, which is exactly what `hashtext` produces.

The majority of the cost of the aggregation is hashing, so it is important that it is fast. 

We could use something like md5 or `sha256`, which have some interesting properties:
- we could generate 64 bit hashes, which will improve the range of cardinalities that we can estimate
- well documented & reproducible in python

However, they produce strings, which need to be cast to integers. This makes the operation considerably slower.


`Hashtext` is undocumented - there have been some questions about whether it should be documented: https://www.postgresql.org/message-id/24463.1329854466%40sss.pgh.pa.us with limited support. Partially because they don't want to guarantee that the current behaviour will be maintained. In that same message, it is suggested that it could be changed to produce 64 bit hashes :)

For our purposes, though, we only care that the results are fast and consistent within the current session.

Reply:
OK just wanted to check that "produce uniformly distributed integers" bit is something we've checked rather than just an assumption. Since it's an internal function it feels like it could be designed to produce unique hashes but not necessarily such that a small change in input results in a big change in the hash